modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+      var nSegmentElems : [localThis.targetLocales.size] int;
+      var locIdx = 0;
+      for loc in localThis.targetLocales do on loc {
+        var nElems = getPrivatizedThis.bag.segments[segmentIdx].nElems.read() : int;
+        nSegmentElems[locIdx] = nElems;
+        locIdx += 1;
+      }
+
+      // Find the average and the excess. The excess is calculated as the amount
+      // of elements a segment has over the average, which is used to calculate
+      // the buffer size for each segment.
+      var total = (+ reduce nSegmentElems);
+      var avg = total / locIdx;
+      var excess : int;
+      for nElems in nSegmentElems {
+        if nElems > avg {
+          excess += nElems - avg;
+        }
+      }
+
+      // Allocate buffer, which holds the 'excess' elements for redistribution.
+      // Then fill it.
+      var buffer = c_malloc(eltType, excess);
+      var bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if nElems > average {
+          var take = nElems - average;
+          var tmpBuffer = buffer + bufferOffset;
+          segment.transferElements(tmpBuffer, take, buffer.locale.id);
+          bufferOffset += take;
+        }
+      }
+
+      // With the excess elements, redistribute it...
+      // BUG: Need to use Chapel Arrays due to issues with compiler automatically
+      // dereferencing a C_Ptr resulting in a segfault.
+      // BUG: Without '--no-denormalization' flag, the compiler will incorrectly
+      // use everything beyond this point by reference and *not* dereference it
+      // beforehand, causing more undefined behavior. For now, I use Chapel Arrays.
+      bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if average > nElems {
+          var give = average - nElems;
+          var arr : [1..give] eltType;
+          on bufferOffset {
+            var tmpBuffer = buffer;
+            for i in 1 .. give {
+              arr[i] = tmpBuffer[bufferOffset];
+              bufferOffset += 1;
+            }
+          }
+          for i in 1 .. give {
+            segment.addElements(arr[i]);
+          }
+        }
+      }
+
+      // Lastly, if there are items left over, just add them to our locale's segment.
+      if excess > bufferOffset {
+        ref segment = localThis.bag.segments[segmentIdx];
+        var give = excess - bufferOffset;
+        var tmpBuffer = buffer + bufferOffset;
+        segment.addElementsPtr(tmpBuffer, give, buffer.locale.id);
+      }
+
+      c_free(buffer);
+    }
+
+    // Phase 3: Release all locks from first node and segment to last node and segment.
+    for loc in localThis.targetLocales do on loc {

```

Comment:
Doesn't this need to happen in the reverse of the acquire order to prevent some deadlock case? Or am I just being paranoid?